crates/provider/src/layers/gas.rs

```diff
@@ -0,0 +1,126 @@
+use crate::{PendingTransactionBuilder, Provider, ProviderLayer, RootProvider};
+use alloy_network::{Network, TransactionBuilder};
+use alloy_primitives::U256;
+use alloy_transport::{Transport, TransportResult};
+use async_trait::async_trait;
+use std::marker::PhantomData;
+
+/// A layer that provides gas estimation for transactions.
+/// Populates the gas price and gas limit for transactions.
+#[derive(Debug, Clone, Copy)]
+pub struct GasEstimatorLayer;
+
+impl<P, N, T> ProviderLayer<P, N, T> for GasEstimatorLayer
+where
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+{
+    type Provider = GasEstimatorProvider<N, T, P>;
+    fn layer(&self, inner: P) -> Self::Provider {
+        GasEstimatorProvider { inner, _phantom: PhantomData }
+    }
+}
+
+/// A provider that provides gas estimation for transactions.
+#[derive(Debug, Clone)]
+pub struct GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    inner: P,
+    _phantom: PhantomData<(N, T)>,
+}
+
+impl<N, T, P> GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    async fn get_gas_price(&self) -> TransportResult<U256> {
+        let gas_price = self.inner.get_gas_price().await?;
+        Ok(gas_price)
+    }
+
+    async fn get_gas_estimate(&self, tx: &N::TransactionRequest) -> TransportResult<U256> {
+        let gas_estimate = self.inner.estimate_gas(tx, None).await?;
+        Ok(gas_estimate)
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<N, T, P> Provider<N, T> for GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    fn root(&self) -> &RootProvider<N, T> {
+        self.inner.root()
+    }
+
+    async fn send_transaction(
+        &self,
+        mut tx: N::TransactionRequest,
+    ) -> TransportResult<PendingTransactionBuilder<'_, N, T>> {
+        if tx.gas_price().is_none() {
+            let gas_price = self.get_gas_price().await?;
+            tx.set_gas_price(gas_price);
+        }
+        if tx.gas_limit().is_none() {
+            let gas_estimate = self.get_gas_estimate(&tx).await?;
+            tx.set_gas_limit(gas_estimate);
+        }
```

Comment:
this should be done with `match` and `join`'ing the 2 futures if both are present, to make it faster.

i also think we should do this for `nonce`, and perhaps this should be a general layer that fills both all fields

Reply:
> i also think we should do this for `nonce`, and perhaps this should be a general layer that fills both all fields
So make generalized tx builder layer and do nonce management and gas estimation in one?

Reply:
Yes I think so cc @prestwich @DaniPopes @mattsse for their view

Reply:
Using `join!` now.